<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Applying Different Thinning Functions • seqgendiff</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Applying Different Thinning Functions" />
<meta property="og:description" content="seqgendiff" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/dcgerard/seqgendiff/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>


<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Applying Different Thinning Functions</h1>
                        <h4 class="author">David Gerard</h4>
            
      
      <small class="dont-index">Source: <a href='http://github.com/dcgerard/seqgendiff/blob/master/vignettes/different_thinners.Rmd'><code>vignettes/different_thinners.Rmd</code></a></small>
      <div class="hidden name"><code>different_thinners.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">library</span>(seqgendiff)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">library</span>(limma)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw">set.seed</span>(<span class="dv">31</span>) <span class="co">## for reproducibility</span></a></code></pre></div>
<div id="abstract" class="section level1">
<h1>Abstract</h1>
<p>We demonstrate how to apply the different thinning functions available in seqgendiff. The following contains guidelines for the workflow of a single repetition in a simulation study.</p>
<p>The methods used here are described in Gerard (2020).</p>
</div>
<div id="generate-data" class="section level1">
<h1>Generate Data</h1>
<p>We will use simulated data for this vignette. Though in practice you would obtain the RNA-seq counts from a real dataset.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">nsamp &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb2-2" title="2">ngene &lt;-<span class="st"> </span><span class="dv">10000</span></a>
<a class="sourceLine" id="cb2-3" title="3">mat   &lt;-<span class="st"> </span><span class="kw">matrix</span>(stats<span class="op">::</span><span class="kw">rpois</span>(<span class="dt">n =</span> nsamp <span class="op">*</span><span class="st"> </span>ngene, <span class="dt">lambda =</span> <span class="dv">100</span>), </a>
<a class="sourceLine" id="cb2-4" title="4">                <span class="dt">nrow =</span> ngene,</a>
<a class="sourceLine" id="cb2-5" title="5">                <span class="dt">ncol =</span> nsamp)</a></code></pre></div>
</div>
<div id="subsetting" class="section level1">
<h1>Subsetting</h1>
<p>Each repetition of a simulation study, you should randomly subset your RNA-seq data so that your results are not dependent on the quirks of a few individuals/genes. The function to do this is <code>select_counts()</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">submat &lt;-<span class="st"> </span><span class="kw">select_counts</span>(<span class="dt">mat =</span> mat, <span class="dt">nsamp =</span> <span class="dv">6</span>, <span class="dt">ngene =</span> <span class="dv">1000</span>)</a></code></pre></div>
<p>The default is to randomly select the samples and genes. Though there are many options available on how to select the genes using the <code>gselect</code> argument.</p>
<p>If the row and column names of the original matrix are <code>NULL</code>, then the row and column names of the returned submatrix contain the indices of the selected rows and columns of the original matrix.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw">head</span>(<span class="kw">rownames</span>(submat))</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="co">#&gt; [1] &quot;23&quot;  &quot;25&quot;  &quot;38&quot;  &quot;39&quot;  &quot;68&quot;  &quot;107&quot;</span></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="kw">head</span>(<span class="kw">colnames</span>(submat))</a>
<a class="sourceLine" id="cb4-4" title="4"><span class="co">#&gt; [1] &quot;18&quot; &quot;44&quot; &quot;62&quot; &quot;69&quot; &quot;92&quot; &quot;99&quot;</span></a></code></pre></div>
</div>
<div id="thinning-library-size" class="section level1">
<h1>Thinning Library Size</h1>
<p>If you are exploring the effects of heterogeneous library sizes, use <code>thin_lib()</code>. You specify the thinning factor on the log-2 scale, so a value of 0 means no thinning, a value of 1 means thin by half, a value of 2 means thin by 1/4, etc. You need to specify this scaling factor for all samples.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">scaling_factor &lt;-<span class="st"> </span><span class="kw">seq_len</span>(<span class="kw">ncol</span>(submat))</a>
<a class="sourceLine" id="cb5-2" title="2">scaling_factor</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co">#&gt; [1] 1 2 3 4 5 6</span></a>
<a class="sourceLine" id="cb5-4" title="4">thout &lt;-<span class="st"> </span><span class="kw">thin_lib</span>(<span class="dt">mat =</span> submat, <span class="dt">thinlog2 =</span> scaling_factor)</a></code></pre></div>
<p>We can verify that thinning was performed correctly by looking at the empirical thinning amount.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="co">## Empirical thinning</span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="kw">colSums</span>(thout<span class="op">$</span>mat) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(submat)</a>
<a class="sourceLine" id="cb6-3" title="3"><span class="co">#&gt;         18         44         62         69         92         99 </span></a>
<a class="sourceLine" id="cb6-4" title="4"><span class="co">#&gt; 0.49887721 0.24906136 0.12624313 0.06288817 0.03108922 0.01598802</span></a>
<a class="sourceLine" id="cb6-5" title="5"></a>
<a class="sourceLine" id="cb6-6" title="6"><span class="co">## Specified thinning</span></a>
<a class="sourceLine" id="cb6-7" title="7"><span class="dv">2</span> <span class="op">^</span><span class="st"> </span><span class="op">-</span>scaling_factor</a>
<a class="sourceLine" id="cb6-8" title="8"><span class="co">#&gt; [1] 0.500000 0.250000 0.125000 0.062500 0.031250 0.015625</span></a></code></pre></div>
<p>A similar function exists to thin gene-wise rather than sample-wise: <code>thin_gene()</code>.</p>
</div>
<div id="total-thinning" class="section level1">
<h1>Total Thinning</h1>
<p>To uniformly thin all counts, use <code>thin_all()</code>. This might be useful for determining read-depth suggestions. It takes as input a single universal scaling factor on the log-2 scale.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">thout &lt;-<span class="st"> </span><span class="kw">thin_all</span>(<span class="dt">mat =</span> submat, <span class="dt">thinlog2 =</span> <span class="dv">1</span>)</a></code></pre></div>
<p>We can verify that we approximately halved all counts:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="kw">sum</span>(thout<span class="op">$</span>mat) <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(submat)</a>
<a class="sourceLine" id="cb8-2" title="2"><span class="co">#&gt; [1] 0.4998192</span></a></code></pre></div>
</div>
<div id="general-thinning" class="section level1">
<h1>General Thinning</h1>
<p>Use <code>thin_diff()</code> for general thinning. For this function, you need to specify both the coefficient matrix and the design matrix.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">designmat &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">each =</span> <span class="kw">ncol</span>(submat) <span class="op">/</span><span class="st"> </span><span class="dv">2</span>), </a>
<a class="sourceLine" id="cb9-2" title="2">                   <span class="kw">rep</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">length.out =</span> <span class="kw">ncol</span>(submat)))</a>
<a class="sourceLine" id="cb9-3" title="3">designmat</a>
<a class="sourceLine" id="cb9-4" title="4"><span class="co">#&gt;      [,1] [,2]</span></a>
<a class="sourceLine" id="cb9-5" title="5"><span class="co">#&gt; [1,]    0    0</span></a>
<a class="sourceLine" id="cb9-6" title="6"><span class="co">#&gt; [2,]    0    1</span></a>
<a class="sourceLine" id="cb9-7" title="7"><span class="co">#&gt; [3,]    0    0</span></a>
<a class="sourceLine" id="cb9-8" title="8"><span class="co">#&gt; [4,]    1    1</span></a>
<a class="sourceLine" id="cb9-9" title="9"><span class="co">#&gt; [5,]    1    0</span></a>
<a class="sourceLine" id="cb9-10" title="10"><span class="co">#&gt; [6,]    1    1</span></a>
<a class="sourceLine" id="cb9-11" title="11"></a>
<a class="sourceLine" id="cb9-12" title="12">coefmat &lt;-<span class="st"> </span><span class="kw">matrix</span>(stats<span class="op">::</span><span class="kw">rnorm</span>(<span class="kw">ncol</span>(designmat) <span class="op">*</span><span class="st"> </span><span class="kw">nrow</span>(submat)),</a>
<a class="sourceLine" id="cb9-13" title="13">                  <span class="dt">ncol =</span> <span class="kw">ncol</span>(designmat),</a>
<a class="sourceLine" id="cb9-14" title="14">                  <span class="dt">nrow =</span> <span class="kw">nrow</span>(submat))</a>
<a class="sourceLine" id="cb9-15" title="15"><span class="kw">head</span>(coefmat)</a>
<a class="sourceLine" id="cb9-16" title="16"><span class="co">#&gt;            [,1]       [,2]</span></a>
<a class="sourceLine" id="cb9-17" title="17"><span class="co">#&gt; [1,]  0.2554242 -1.3119342</span></a>
<a class="sourceLine" id="cb9-18" title="18"><span class="co">#&gt; [2,]  0.6377614  0.5336341</span></a>
<a class="sourceLine" id="cb9-19" title="19"><span class="co">#&gt; [3,]  1.3052848  0.6763963</span></a>
<a class="sourceLine" id="cb9-20" title="20"><span class="co">#&gt; [4,]  0.1383671 -0.9041101</span></a>
<a class="sourceLine" id="cb9-21" title="21"><span class="co">#&gt; [5,] -0.1251100  1.3601036</span></a>
<a class="sourceLine" id="cb9-22" title="22"><span class="co">#&gt; [6,] -0.8867430 -0.5249070</span></a></code></pre></div>
<p>Once we have the coefficient and design matrices, we can thin.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">thout &lt;-<span class="st"> </span><span class="kw">thin_diff</span>(<span class="dt">mat          =</span> submat, </a>
<a class="sourceLine" id="cb10-2" title="2">                   <span class="dt">design_fixed =</span> designmat, </a>
<a class="sourceLine" id="cb10-3" title="3">                   <span class="dt">coef_fixed   =</span> coefmat)</a></code></pre></div>
<p>We can verify that we thinned correctly using the voom-limma pipeline.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">new_design &lt;-<span class="st"> </span><span class="kw">cbind</span>(thout<span class="op">$</span>design_obs, thout<span class="op">$</span>designmat)</a>
<a class="sourceLine" id="cb11-2" title="2">vout &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">voom</span>(<span class="dt">counts =</span> thout<span class="op">$</span>mat, <span class="dt">design =</span> new_design)</a>
<a class="sourceLine" id="cb11-3" title="3">lout &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">lmFit</span>(vout)</a>
<a class="sourceLine" id="cb11-4" title="4">coefhat &lt;-<span class="st"> </span><span class="kw">coef</span>(lout)[, <span class="dv">-1</span>, drop =<span class="st"> </span><span class="ot">FALSE</span>]</a></code></pre></div>
<p>We’ll plot the true coefficients against their estimates.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">oldpar &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="fl">2.5</span>, <span class="fl">2.5</span>, <span class="dv">1</span>, <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span><span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb12-2" title="2">              <span class="dt">mgp =</span> <span class="kw">c</span>(<span class="fl">1.5</span>, <span class="fl">0.5</span>, <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb12-3" title="3"><span class="kw">plot</span>(<span class="dt">x    =</span> coefmat[, <span class="dv">1</span>], </a>
<a class="sourceLine" id="cb12-4" title="4">     <span class="dt">y    =</span> coefhat[, <span class="dv">1</span>], </a>
<a class="sourceLine" id="cb12-5" title="5">     <span class="dt">xlab =</span> <span class="st">&quot;True Coefficient&quot;</span>, </a>
<a class="sourceLine" id="cb12-6" title="6">     <span class="dt">ylab =</span> <span class="st">&quot;Estimated Coefficient&quot;</span>,</a>
<a class="sourceLine" id="cb12-7" title="7">     <span class="dt">main =</span> <span class="st">&quot;First Variable&quot;</span>,</a>
<a class="sourceLine" id="cb12-8" title="8">     <span class="dt">pch  =</span> <span class="dv">16</span>)</a>
<a class="sourceLine" id="cb12-9" title="9"><span class="kw">abline</span>(<span class="dt">a   =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb12-10" title="10">       <span class="dt">b   =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb12-11" title="11">       <span class="dt">lty =</span> <span class="dv">2</span>, </a>
<a class="sourceLine" id="cb12-12" title="12">       <span class="dt">col =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb12-13" title="13">       <span class="dt">lwd =</span> <span class="dv">2</span>)</a></code></pre></div>
<p><img src="C:/Users/dgerard/Dropbox/seqgendiff/docs/articles/different_thinners_files/figure-html/unnamed-chunk-12-1.png" width="700" /></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"></a>
<a class="sourceLine" id="cb13-2" title="2"><span class="kw">plot</span>(<span class="dt">x    =</span> coefmat[, <span class="dv">2</span>], </a>
<a class="sourceLine" id="cb13-3" title="3">     <span class="dt">y    =</span> coefhat[, <span class="dv">2</span>], </a>
<a class="sourceLine" id="cb13-4" title="4">     <span class="dt">xlab =</span> <span class="st">&quot;True Coefficient&quot;</span>, </a>
<a class="sourceLine" id="cb13-5" title="5">     <span class="dt">ylab =</span> <span class="st">&quot;Estimated Coefficient&quot;</span>,</a>
<a class="sourceLine" id="cb13-6" title="6">     <span class="dt">main =</span> <span class="st">&quot;Second Variable&quot;</span>,</a>
<a class="sourceLine" id="cb13-7" title="7">     <span class="dt">pch  =</span> <span class="dv">16</span>)</a>
<a class="sourceLine" id="cb13-8" title="8"><span class="kw">abline</span>(<span class="dt">a   =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb13-9" title="9">       <span class="dt">b   =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb13-10" title="10">       <span class="dt">lty =</span> <span class="dv">2</span>, </a>
<a class="sourceLine" id="cb13-11" title="11">       <span class="dt">col =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb13-12" title="12">       <span class="dt">lwd =</span> <span class="dv">2</span>)</a></code></pre></div>
<p><img src="C:/Users/dgerard/Dropbox/seqgendiff/docs/articles/different_thinners_files/figure-html/unnamed-chunk-12-2.png" width="700" /></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="kw">par</span>(oldpar)</a></code></pre></div>
</div>
<div id="correlation-with-surrogate-variables" class="section level1">
<h1>Correlation with Surrogate Variables</h1>
<p>The difference between the <code>design_fixed</code> and <code>design_perm</code> arguments is that the rows in <code>design_perm</code> are permuted before applying thinning. Without any other arguments, this makes the design variables independent of any surrogate variables. With the additional specification of the <code>target_cor</code> argument, we try to control the amount of correlation between the design variables in <code>design_perm</code> and any surrogate variables.</p>
<p>Let’s target for a correlation of 0.9 between the first surrogate variable and the first design variable, and a correlation of 0 between the first surrogate variable and the second design variable.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">target_cor &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="fl">0.9</span>, <span class="dv">0</span>), <span class="dt">nrow =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb15-2" title="2">target_cor</a>
<a class="sourceLine" id="cb15-3" title="3"><span class="co">#&gt;      [,1]</span></a>
<a class="sourceLine" id="cb15-4" title="4"><span class="co">#&gt; [1,]  0.9</span></a>
<a class="sourceLine" id="cb15-5" title="5"><span class="co">#&gt; [2,]  0.0</span></a>
<a class="sourceLine" id="cb15-6" title="6">thout_cor &lt;-<span class="st"> </span><span class="kw">thin_diff</span>(<span class="dt">mat         =</span> submat,</a>
<a class="sourceLine" id="cb15-7" title="7">                       <span class="dt">design_perm =</span> designmat, </a>
<a class="sourceLine" id="cb15-8" title="8">                       <span class="dt">coef_perm   =</span> coefmat, </a>
<a class="sourceLine" id="cb15-9" title="9">                       <span class="dt">target_cor  =</span> target_cor)</a>
<a class="sourceLine" id="cb15-10" title="10"><span class="co">#&gt; Note that optmatch uses a strange non-standard license:</span></a>
<a class="sourceLine" id="cb15-11" title="11"><span class="co">#&gt; https://cran.r-project.org/package=optmatch/LICENSE</span></a>
<a class="sourceLine" id="cb15-12" title="12"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb15-13" title="13"><span class="co">#&gt; If this doesn&#39;t work for you, try permute_method = &quot;hungarian&quot;</span></a>
<a class="sourceLine" id="cb15-14" title="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb15-15" title="15"><span class="co">#&gt; This message is displayed once per R session.</span></a></code></pre></div>
<p>The first variable is indeed more strongly correlated with the estimated surrogate variable:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="kw">cor</span>(thout_cor<span class="op">$</span>designmat, thout_cor<span class="op">$</span>sv)</a>
<a class="sourceLine" id="cb16-2" title="2"><span class="co">#&gt;         [,1]</span></a>
<a class="sourceLine" id="cb16-3" title="3"><span class="co">#&gt; P1 0.7473094</span></a>
<a class="sourceLine" id="cb16-4" title="4"><span class="co">#&gt; P2 0.2976213</span></a></code></pre></div>
</div>
<div id="estimate-actual-correlation" class="section level1">
<h1>Estimate Actual Correlation</h1>
<p>The actual correlation between the permuted design matrix and the surrogate variables will not be the target correlation. But we can estimate what the actual correlation is using the function <code>effective_cor()</code>.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">eout &lt;-<span class="st"> </span><span class="kw">effective_cor</span>(<span class="dt">design_perm =</span> designmat, </a>
<a class="sourceLine" id="cb17-2" title="2">                      <span class="dt">sv          =</span> thout_cor<span class="op">$</span>sv, </a>
<a class="sourceLine" id="cb17-3" title="3">                      <span class="dt">target_cor  =</span> target_cor, </a>
<a class="sourceLine" id="cb17-4" title="4">                      <span class="dt">iternum     =</span> <span class="dv">50</span>)</a>
<a class="sourceLine" id="cb17-5" title="5">eout</a>
<a class="sourceLine" id="cb17-6" title="6"><span class="co">#&gt;           [,1]</span></a>
<a class="sourceLine" id="cb17-7" title="7"><span class="co">#&gt; [1,] 0.7113003</span></a>
<a class="sourceLine" id="cb17-8" title="8"><span class="co">#&gt; [2,] 0.1399246</span></a></code></pre></div>
<p>I am only using 50 iterations here for speed reasons, but you should stick to the defaults for <code>iternum</code>.</p>
</div>
<div id="special-case-of-two-group-model" class="section level1">
<h1>Special Case of Two-group Model</h1>
<p>For the special case when your design matrix is just a group indicator (that is, you have two groups of individuals), you can use the function <code>thin_2group()</code>. Let’s generate data from the two-group model where 90% of genes are null and the non-null effects are gamma-distributed.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1">thout &lt;-<span class="st"> </span><span class="kw">thin_2group</span>(<span class="dt">mat           =</span> submat, </a>
<a class="sourceLine" id="cb18-2" title="2">                     <span class="dt">prop_null     =</span> <span class="fl">0.9</span>, </a>
<a class="sourceLine" id="cb18-3" title="3">                     <span class="dt">signal_fun    =</span> stats<span class="op">::</span>rgamma,</a>
<a class="sourceLine" id="cb18-4" title="4">                     <span class="dt">signal_params =</span> <span class="kw">list</span>(<span class="dt">shape =</span> <span class="dv">1</span>, <span class="dt">rate =</span> <span class="dv">1</span>))</a></code></pre></div>
<p>We can again verify that we thinned appropriately using the voom-limma pipeline:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">new_design &lt;-<span class="st"> </span><span class="kw">cbind</span>(thout<span class="op">$</span>design_obs, thout<span class="op">$</span>designmat)</a>
<a class="sourceLine" id="cb19-2" title="2">new_design</a>
<a class="sourceLine" id="cb19-3" title="3"><span class="co">#&gt;      (Intercept) P1</span></a>
<a class="sourceLine" id="cb19-4" title="4"><span class="co">#&gt; [1,]           1  0</span></a>
<a class="sourceLine" id="cb19-5" title="5"><span class="co">#&gt; [2,]           1  0</span></a>
<a class="sourceLine" id="cb19-6" title="6"><span class="co">#&gt; [3,]           1  0</span></a>
<a class="sourceLine" id="cb19-7" title="7"><span class="co">#&gt; [4,]           1  1</span></a>
<a class="sourceLine" id="cb19-8" title="8"><span class="co">#&gt; [5,]           1  1</span></a>
<a class="sourceLine" id="cb19-9" title="9"><span class="co">#&gt; [6,]           1  1</span></a>
<a class="sourceLine" id="cb19-10" title="10">vout &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">voom</span>(<span class="dt">counts =</span> thout<span class="op">$</span>mat, <span class="dt">design =</span> new_design)</a>
<a class="sourceLine" id="cb19-11" title="11">lout &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">lmFit</span>(vout)</a>
<a class="sourceLine" id="cb19-12" title="12">coefhat &lt;-<span class="st"> </span><span class="kw">coef</span>(lout)[, <span class="dv">2</span>, drop =<span class="st"> </span><span class="ot">FALSE</span>]</a></code></pre></div>
<p>And we can plot the results</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1">oldpar &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="fl">2.5</span>, <span class="fl">2.5</span>, <span class="dv">1</span>, <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span><span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb20-2" title="2">              <span class="dt">mgp =</span> <span class="kw">c</span>(<span class="fl">1.5</span>, <span class="fl">0.5</span>, <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb20-3" title="3"><span class="kw">plot</span>(<span class="dt">x    =</span> thout<span class="op">$</span>coefmat, </a>
<a class="sourceLine" id="cb20-4" title="4">     <span class="dt">y    =</span> coefhat, </a>
<a class="sourceLine" id="cb20-5" title="5">     <span class="dt">xlab =</span> <span class="st">&quot;True Coefficient&quot;</span>, </a>
<a class="sourceLine" id="cb20-6" title="6">     <span class="dt">ylab =</span> <span class="st">&quot;Estimated Coefficient&quot;</span>,</a>
<a class="sourceLine" id="cb20-7" title="7">     <span class="dt">main =</span> <span class="st">&quot;First Variable&quot;</span>,</a>
<a class="sourceLine" id="cb20-8" title="8">     <span class="dt">pch  =</span> <span class="dv">16</span>)</a>
<a class="sourceLine" id="cb20-9" title="9"><span class="kw">abline</span>(<span class="dt">a   =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb20-10" title="10">       <span class="dt">b   =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb20-11" title="11">       <span class="dt">lty =</span> <span class="dv">2</span>, </a>
<a class="sourceLine" id="cb20-12" title="12">       <span class="dt">col =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb20-13" title="13">       <span class="dt">lwd =</span> <span class="dv">2</span>)</a></code></pre></div>
<p><img src="C:/Users/dgerard/Dropbox/seqgendiff/docs/articles/different_thinners_files/figure-html/unnamed-chunk-18-1.png" width="700" /></p>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<ul>
<li>Gerard, D (2020). “Data-based RNA-seq simulations by binomial thinning.” <em>BMC Bioinformatics</em>. 21(1), 206. doi: <a href="https://doi.org/10.1186/s12859-020-3450-9">10.1186/s12859-020-3450-9</a>.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc">
      <h2 data-toc-skip>Contents</h2>
    </nav>
      </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


