<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Simulate RNA-seq Data from Real Data • seqgendiff</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Simulate RNA-seq Data from Real Data" />
<meta property="og:description" content="seqgendiff" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/dcgerard/seqgendiff/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>


<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simulate RNA-seq Data from Real Data</h1>
                        <h4 class="author">David Gerard</h4>
            
      
      <small class="dont-index">Source: <a href='http://github.com/dcgerard/seqgendiff/blob/master/vignettes/simulate_rnaseq.Rmd'><code>vignettes/simulate_rnaseq.Rmd</code></a></small>
      <div class="hidden name"><code>simulate_rnaseq.Rmd</code></div>

    </div>

    
    
<div id="abstract" class="section level1">
<h1>Abstract</h1>
<p>We demonstrate how one may use seqgendiff in differential expression simulation studies using the airway data from Himes et al (2014). We use seqgendiff to simulate one dataset which we then analyze with two pipelines: the sva-voom-limma-eBayes-qvalue pipeline, and the sva-DESeq2-qvalue pipeline. In practice, you would simulate many datasets and compare average performance. But playing with a single dataset can help you gain intuition with various methods.</p>
<p>The methods used here are described in Gerard (2020).</p>
</div>
<div id="analysis" class="section level1">
<h1>Analysis</h1>
<p>Set the seed for reproducibility:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">set.seed</span>(<span class="dv">1</span>)</a></code></pre></div>
<p>The airway data are available in the airway R package from Bioconductor. If you don’t have this package, you can install it with the BiocManager package:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb2-2" title="2">    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb2-3" title="3"></a>
<a class="sourceLine" id="cb2-4" title="4">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;airway&quot;</span>)</a></code></pre></div>
<p>Now we can load these data into R.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="kw">library</span>(airway)</a>
<a class="sourceLine" id="cb3-2" title="2"><span class="kw">data</span>(<span class="st">&quot;airway&quot;</span>)</a></code></pre></div>
<p>You can read about these data <a href="https://bioconductor.org/packages/airway/">here</a>.</p>
<p>These are the other packages that we’ll need for this vignette:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw">library</span>(seqgendiff)</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="kw">library</span>(SummarizedExperiment)</a>
<a class="sourceLine" id="cb4-3" title="3"><span class="kw">library</span>(DESeq2)</a>
<a class="sourceLine" id="cb4-4" title="4"><span class="kw">library</span>(limma)</a>
<a class="sourceLine" id="cb4-5" title="5"><span class="kw">library</span>(sva)</a>
<a class="sourceLine" id="cb4-6" title="6"><span class="kw">library</span>(qvalue)</a></code></pre></div>
<p>The last five packages are all from Bioconductor, which you need to install using <code>BiocManager::install()</code>.</p>
<p>The airway dataset comes with a few observed variables which affect gene expression. We’ll consider these “surrogate variables”. That is, we’ll add signal to the data and try to account for these surrogate variables.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">coldat &lt;-<span class="st"> </span><span class="kw">colData</span>(airway)[, <span class="kw">c</span>(<span class="st">&quot;cell&quot;</span>, <span class="st">&quot;dex&quot;</span>)]</a>
<a class="sourceLine" id="cb5-2" title="2">true_sv &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span>cell <span class="op">+</span><span class="st"> </span>dex, <span class="dt">data =</span> coldat)[, <span class="dv">-1</span>]</a>
<a class="sourceLine" id="cb5-3" title="3">true_sv</a>
<a class="sourceLine" id="cb5-4" title="4"><span class="co">#&gt;            cellN061011 cellN080611 cellN61311 dexuntrt</span></a>
<a class="sourceLine" id="cb5-5" title="5"><span class="co">#&gt; SRR1039508           0           0          1        1</span></a>
<a class="sourceLine" id="cb5-6" title="6"><span class="co">#&gt; SRR1039509           0           0          1        0</span></a>
<a class="sourceLine" id="cb5-7" title="7"><span class="co">#&gt; SRR1039512           0           0          0        1</span></a>
<a class="sourceLine" id="cb5-8" title="8"><span class="co">#&gt; SRR1039513           0           0          0        0</span></a>
<a class="sourceLine" id="cb5-9" title="9"><span class="co">#&gt; SRR1039516           0           1          0        1</span></a>
<a class="sourceLine" id="cb5-10" title="10"><span class="co">#&gt; SRR1039517           0           1          0        0</span></a>
<a class="sourceLine" id="cb5-11" title="11"><span class="co">#&gt; SRR1039520           1           0          0        1</span></a>
<a class="sourceLine" id="cb5-12" title="12"><span class="co">#&gt; SRR1039521           1           0          0        0</span></a></code></pre></div>
<p>Suppose we want to add a treatment effect to the RNA-seq data. We’ll first remove all genes with no counts:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">allzero &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">assay</span>(airway)) <span class="op">&lt;</span><span class="st"> </span><span class="dv">10</span><span class="op">^-</span><span class="dv">6</span></a>
<a class="sourceLine" id="cb6-2" title="2">airway &lt;-<span class="st"> </span>airway[<span class="op">!</span>allzero, ]</a></code></pre></div>
<p>Then we’ll use the convenient <code>thin_2group()</code> function from the seqgendiff package to add a <span class="math inline">\(N(0, 0.8^2)\)</span> log2-effect size to 10% of the genes:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">thout &lt;-<span class="st"> </span><span class="kw">thin_2group</span>(<span class="dt">mat =</span> <span class="kw">assay</span>(airway), </a>
<a class="sourceLine" id="cb7-2" title="2">                     <span class="dt">prop_null =</span> <span class="fl">0.9</span>, </a>
<a class="sourceLine" id="cb7-3" title="3">                     <span class="dt">signal_fun =</span> stats<span class="op">::</span>rnorm,</a>
<a class="sourceLine" id="cb7-4" title="4">                     <span class="dt">signal_params =</span> <span class="kw">list</span>(<span class="dt">mean =</span> <span class="dv">0</span>, <span class="dt">sd =</span> <span class="fl">0.8</span>))</a></code></pre></div>
<p>Let’s see how well the sva-voom-limma-ebayes-qvalue pipeline does estimating these effects. First, we’ll fit this pipeline.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">X &lt;-<span class="st"> </span><span class="kw">cbind</span>(thout<span class="op">$</span>design_obs, thout<span class="op">$</span>designmat)</a>
<a class="sourceLine" id="cb8-2" title="2">Y &lt;-<span class="st"> </span><span class="kw">log2</span>(thout<span class="op">$</span>mat <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span>)</a>
<a class="sourceLine" id="cb8-3" title="3">n_sv &lt;-<span class="st"> </span><span class="kw">num.sv</span>(<span class="dt">dat =</span> Y, <span class="dt">mod =</span> X)</a>
<a class="sourceLine" id="cb8-4" title="4">svout &lt;-<span class="st"> </span><span class="kw">sva</span>(<span class="dt">dat =</span> Y, <span class="dt">mod =</span> X, <span class="dt">n.sv =</span> n_sv)</a>
<a class="sourceLine" id="cb8-5" title="5"><span class="co">#&gt; Number of significant surrogate variables is:  2 </span></a>
<a class="sourceLine" id="cb8-6" title="6"><span class="co">#&gt; Iteration (out of 5 ):1  2  3  4  5</span></a>
<a class="sourceLine" id="cb8-7" title="7">vout &lt;-<span class="st"> </span><span class="kw">voom</span>(<span class="dt">counts =</span> thout<span class="op">$</span>mat, <span class="dt">design =</span> <span class="kw">cbind</span>(X, svout<span class="op">$</span>sv))</a>
<a class="sourceLine" id="cb8-8" title="8">lout &lt;-<span class="st"> </span><span class="kw">lmFit</span>(vout)</a>
<a class="sourceLine" id="cb8-9" title="9">eout &lt;-<span class="st"> </span><span class="kw">eBayes</span>(lout)</a>
<a class="sourceLine" id="cb8-10" title="10">qout &lt;-<span class="st"> </span><span class="kw">qvalue</span>(<span class="dt">p =</span> eout<span class="op">$</span>p.value[, <span class="dv">2</span>])</a>
<a class="sourceLine" id="cb8-11" title="11">bhat &lt;-<span class="st"> </span>eout<span class="op">$</span>coefficients[, <span class="dv">2</span>]</a></code></pre></div>
<p>We plot the coefficient estimates against their true values.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="kw">plot</span>(thout<span class="op">$</span>coefmat, </a>
<a class="sourceLine" id="cb9-2" title="2">     bhat, </a>
<a class="sourceLine" id="cb9-3" title="3">     <span class="dt">xlab =</span> <span class="st">&quot;True Coefficients&quot;</span>, </a>
<a class="sourceLine" id="cb9-4" title="4">     <span class="dt">ylab =</span> <span class="st">&quot;Estimated Coefficients&quot;</span>)</a>
<a class="sourceLine" id="cb9-5" title="5"><span class="kw">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dt">col =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> <span class="dv">2</span>)</a></code></pre></div>
<p><img src="C:/Users/dgerard/Dropbox/seqgendiff/docs/articles/simulate_rnaseq_files/figure-html/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We plot the q-values against the null status of each gene.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">is_null_gene &lt;-<span class="st"> </span><span class="kw">abs</span>(thout<span class="op">$</span>coefmat) <span class="op">&lt;</span><span class="st"> </span><span class="dv">10</span><span class="op">^-</span><span class="dv">6</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="kw">boxplot</span>(qout<span class="op">$</span>qvalues <span class="op">~</span><span class="st"> </span>is_null_gene,</a>
<a class="sourceLine" id="cb10-3" title="3">        <span class="dt">xlab =</span> <span class="st">&quot;Null Gene&quot;</span>,</a>
<a class="sourceLine" id="cb10-4" title="4">        <span class="dt">ylab =</span> <span class="st">&quot;q-value&quot;</span>)</a></code></pre></div>
<p><img src="C:/Users/dgerard/Dropbox/seqgendiff/docs/articles/simulate_rnaseq_files/figure-html/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Does the sva-DESeq2 pipeline do any better? We’ll first use <code>seqgendiff::ThinDataToDESeqDataSet()</code> to convert the <code>ThinData</code> object <code>thout</code> into a <code>DESeqDataSet</code> object.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">thout<span class="op">$</span>design_obs &lt;-<span class="st"> </span><span class="kw">cbind</span>(thout<span class="op">$</span>design_obs, svout<span class="op">$</span>sv)</a>
<a class="sourceLine" id="cb11-2" title="2">dds &lt;-<span class="st"> </span><span class="kw">ThinDataToDESeqDataSet</span>(<span class="dt">obj =</span> thout)</a>
<a class="sourceLine" id="cb11-3" title="3"><span class="kw">colData</span>(dds)</a>
<a class="sourceLine" id="cb11-4" title="4"><span class="co">#&gt; DataFrame with 8 rows and 3 columns</span></a>
<a class="sourceLine" id="cb11-5" title="5"><span class="co">#&gt;                 V1         V2        P1</span></a>
<a class="sourceLine" id="cb11-6" title="6"><span class="co">#&gt;          &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb11-7" title="7"><span class="co">#&gt; sample1  0.0484565 -0.4083835         1</span></a>
<a class="sourceLine" id="cb11-8" title="8"><span class="co">#&gt; sample2 -0.2708569  0.0965681         0</span></a>
<a class="sourceLine" id="cb11-9" title="9"><span class="co">#&gt; sample3  0.2242451 -0.3339131         0</span></a>
<a class="sourceLine" id="cb11-10" title="10"><span class="co">#&gt; sample4 -0.5995780  0.1820605         1</span></a>
<a class="sourceLine" id="cb11-11" title="11"><span class="co">#&gt; sample5  0.4572693 -0.0471994         0</span></a>
<a class="sourceLine" id="cb11-12" title="12"><span class="co">#&gt; sample6  0.4676617  0.6664741         1</span></a>
<a class="sourceLine" id="cb11-13" title="13"><span class="co">#&gt; sample7 -0.0348024 -0.4100166         0</span></a>
<a class="sourceLine" id="cb11-14" title="14"><span class="co">#&gt; sample8 -0.2923952  0.2544100         1</span></a>
<a class="sourceLine" id="cb11-15" title="15"><span class="kw">design</span>(dds)</a>
<a class="sourceLine" id="cb11-16" title="16"><span class="co">#&gt; ~V1 + V2 + P1</span></a>
<a class="sourceLine" id="cb11-17" title="17"><span class="co">#&gt; &lt;environment: 0x000000003037cb78&gt;</span></a></code></pre></div>
<p>Note that we added the estimated surrogate variables from sva to the <code>ThinData</code> object before converting to a <code>DESeqDataSet</code> object.</p>
<p>Now we will fit DESeq2.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">dds &lt;-<span class="st"> </span><span class="kw">DESeq</span>(<span class="dt">object =</span> dds)</a>
<a class="sourceLine" id="cb12-2" title="2"><span class="co">#&gt; estimating size factors</span></a>
<a class="sourceLine" id="cb12-3" title="3"><span class="co">#&gt; estimating dispersions</span></a>
<a class="sourceLine" id="cb12-4" title="4"><span class="co">#&gt; gene-wise dispersion estimates</span></a>
<a class="sourceLine" id="cb12-5" title="5"><span class="co">#&gt; mean-dispersion relationship</span></a>
<a class="sourceLine" id="cb12-6" title="6"><span class="co">#&gt; final dispersion estimates</span></a>
<a class="sourceLine" id="cb12-7" title="7"><span class="co">#&gt; fitting model and testing</span></a>
<a class="sourceLine" id="cb12-8" title="8">pval_dds &lt;-<span class="st"> </span><span class="kw">rowData</span>(dds)[, <span class="st">&quot;WaldPvalue_P1&quot;</span>]</a>
<a class="sourceLine" id="cb12-9" title="9">qval_dds &lt;-<span class="st"> </span><span class="kw">qvalue</span>(<span class="dt">p =</span> pval_dds)</a></code></pre></div>
<p>We plot the coefficient estimates against their true values.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">coefmat &lt;-<span class="st"> </span><span class="kw">rowData</span>(dds)[, <span class="kw">c</span>(<span class="st">&quot;true_P1&quot;</span>, <span class="st">&quot;P1&quot;</span>)]</a>
<a class="sourceLine" id="cb13-2" title="2"><span class="kw">plot</span>(coefmat<span class="op">$</span>true_P1, </a>
<a class="sourceLine" id="cb13-3" title="3">     coefmat<span class="op">$</span>P1,</a>
<a class="sourceLine" id="cb13-4" title="4">     <span class="dt">xlab =</span> <span class="st">&quot;True Coefficients&quot;</span>, </a>
<a class="sourceLine" id="cb13-5" title="5">     <span class="dt">ylab =</span> <span class="st">&quot;Estimated Coefficients&quot;</span>)</a>
<a class="sourceLine" id="cb13-6" title="6"><span class="kw">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dt">col =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> <span class="dv">2</span>)</a></code></pre></div>
<p><img src="C:/Users/dgerard/Dropbox/seqgendiff/docs/articles/simulate_rnaseq_files/figure-html/unnamed-chunk-36-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We also plot the q-values against the null status of each gene.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="kw">boxplot</span>(qval_dds<span class="op">$</span>qvalues <span class="op">~</span><span class="st"> </span>is_null_gene,</a>
<a class="sourceLine" id="cb14-2" title="2">        <span class="dt">xlab =</span> <span class="st">&quot;Null Gene&quot;</span>,</a>
<a class="sourceLine" id="cb14-3" title="3">        <span class="dt">ylab =</span> <span class="st">&quot;q-value&quot;</span>)</a></code></pre></div>
<p><img src="C:/Users/dgerard/Dropbox/seqgendiff/docs/articles/simulate_rnaseq_files/figure-html/unnamed-chunk-38-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The MSE is worse for DESeq2.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">mse_limma &lt;-<span class="st"> </span><span class="kw">mean</span>((bhat <span class="op">-</span><span class="st"> </span>thout<span class="op">$</span>coefmat) <span class="op">^</span><span class="st"> </span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb15-2" title="2">mse_deseq2 &lt;-<span class="st"> </span><span class="kw">mean</span>((coefmat<span class="op">$</span>true_P1 <span class="op">-</span><span class="st"> </span>coefmat<span class="op">$</span>P1) <span class="op">^</span><span class="st"> </span><span class="dv">2</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb15-3" title="3">mse_deseq2</a>
<a class="sourceLine" id="cb15-4" title="4"><span class="co">#&gt; [1] 1.095837</span></a>
<a class="sourceLine" id="cb15-5" title="5">mse_limma</a>
<a class="sourceLine" id="cb15-6" title="6"><span class="co">#&gt; [1] 0.2910926</span></a></code></pre></div>
<p>The FDP is about the same for the two pipelines. Both are OK, but conservative.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="kw">mean</span>(is_null_gene[qval_dds<span class="op">$</span>qvalues <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.1</span>], <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb16-2" title="2"><span class="co">#&gt; [1] 0.05263158</span></a>
<a class="sourceLine" id="cb16-3" title="3"><span class="kw">mean</span>(is_null_gene[qout<span class="op">$</span>qvalues <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.1</span>], <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb16-4" title="4"><span class="co">#&gt; [1] 0.01336898</span></a></code></pre></div>
<p>DESeq2 has more discoveries</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="kw">sum</span>(qval_dds<span class="op">$</span>qvalues <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.1</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb17-2" title="2"><span class="co">#&gt; [1] 513</span></a>
<a class="sourceLine" id="cb17-3" title="3"><span class="kw">sum</span>(qout<span class="op">$</span>qvalues <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.1</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb17-4" title="4"><span class="co">#&gt; [1] 374</span></a></code></pre></div>
<p>As a side-note, it seems that the second surrogate variable is getting at the dexamethasone variable.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="kw">boxplot</span>(svout<span class="op">$</span>sv[, <span class="dv">2</span>] <span class="op">~</span><span class="st"> </span>true_sv[, <span class="dv">4</span>], </a>
<a class="sourceLine" id="cb18-2" title="2">        <span class="dt">xlab =</span> <span class="st">&quot;Dexamethasone Treatment&quot;</span>, </a>
<a class="sourceLine" id="cb18-3" title="3">        <span class="dt">ylab =</span> <span class="st">&quot;2nd Surrogate Variable&quot;</span>)</a>
<a class="sourceLine" id="cb18-4" title="4"><span class="kw">points</span>(<span class="kw">jitter</span>(true_sv[, <span class="dv">4</span>] <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), svout<span class="op">$</span>sv[, <span class="dv">2</span>])</a></code></pre></div>
<p><img src="C:/Users/dgerard/Dropbox/seqgendiff/docs/articles/simulate_rnaseq_files/figure-html/unnamed-chunk-46-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<ul>
<li>Gerard, D (2020). “Data-based RNA-seq simulations by binomial thinning.” <em>BMC Bioinformatics</em>. 21(1), 206. doi: <a href="https://doi.org/10.1186/s12859-020-3450-9">10.1186/s12859-020-3450-9</a>.</li>
<li>Himes, Blanca E., Xiaofeng Jiang, Peter Wagner, Ruoxi Hu, Qiyu Wang, Barbara Klanderman, Reid M. Whitaker et al. “RNA-Seq transcriptome profiling identifies CRISPLD2 as a glucocorticoid responsive gene that modulates cytokine function in airway smooth muscle cells.” <em>PloS one</em> 9, no. 6 (2014): e99625.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc">
      <h2 data-toc-skip>Contents</h2>
    </nav>
      </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


