<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Apply Poisson thinning to a matrix of count data. — poisthin • seqgendiff</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Apply Poisson thinning to a matrix of count data. — poisthin" />

<meta property="og:description" content="This is now defunct. Please try out select_counts and
thin_2group." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/seqgendiff">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Apply Poisson thinning to a matrix of count data.</h1>
    <small class="dont-index">Source: <a href='https://github.com/dcgerard/seqgendiff/blob/master/R/poisthin.R'><code>R/poisthin.R</code></a></small>
    <div class="hidden name"><code>poisthin.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This is now defunct. Please try out <code><a href='select_counts.html'>select_counts</a></code> and
<code><a href='thin_2group.html'>thin_2group</a></code>.</p>
    
    </div>

    <pre class="usage"><span class='fu'>poisthin</span>(<span class='no'>mat</span>, <span class='kw'>nsamp</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/nrow'>nrow</a></span>(<span class='no'>mat</span>), <span class='kw'>ngene</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/nrow'>ncol</a></span>(<span class='no'>mat</span>),
  <span class='kw'>gselect</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"max"</span>, <span class='st'>"random"</span>, <span class='st'>"rand_max"</span>, <span class='st'>"custom"</span>, <span class='st'>"mean_max"</span>),
  <span class='kw'>gvec</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>skip_gene</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>signal_fun</span> <span class='kw'>=</span> <span class='kw pkg'>stats</span><span class='kw ns'>::</span><span class='no'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>,
  <span class='kw'>signal_params</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/list'>list</a></span>(<span class='kw'>mean</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>sd</span> <span class='kw'>=</span> <span class='fl'>1</span>), <span class='kw'>prop_null</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>group_assign</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"frac"</span>, <span class='st'>"random"</span>, <span class='st'>"cor"</span>), <span class='kw'>group_prop</span> <span class='kw'>=</span> <span class='fl'>0.5</span>,
  <span class='kw'>corvec</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mat</th>
      <td><p>A matrix of count data. The rows index the individuals and
the columns index the genes.</p></td>
    </tr>
    <tr>
      <th>nsamp</th>
      <td><p>The number of samples to select from <code>mat</code>.</p></td>
    </tr>
    <tr>
      <th>ngene</th>
      <td><p>The number of genes to select from <code>mat</code>.</p></td>
    </tr>
    <tr>
      <th>gselect</th>
      <td><p>How should we select the subset of genes? Should we choose
the <code>ngene</code> most median expressed genes (<code>"max"</code>), a random sample
of the genes (<code>"random"</code>), a random sample of the most expressed
genes (<code>"rand_max"</code>), a user-provided list (<code>"custom"</code>), or by maximum
mean expression level (<code>"mean_max"</code>)?
If <code>"custom"</code>, then <code>gvec</code> should be specified. Expression levels
of a gene are measured by median expression across individuals with ties broken
by mean expression.</p></td>
    </tr>
    <tr>
      <th>gvec</th>
      <td><p>A logical of length <code><a href='https://www.rdocumentation.org/packages/base/topics/nrow'>ncol(mat)</a></code>. A <code>TRUE</code> in position \(i\)
indicates inclusion into the smaller dataset. Hence, <code><a href='https://www.rdocumentation.org/packages/base/topics/sum'>sum(gvec)</a></code> should
equal <code>ngene</code>.</p></td>
    </tr>
    <tr>
      <th>skip_gene</th>
      <td><p>The number of maximally expressed genes to skip.
Not used if <code>gselect = "custom"</code>.</p></td>
    </tr>
    <tr>
      <th>signal_fun</th>
      <td><p>A function that returns the signal. This should take as
input <code>n</code> for the number of samples to return and then return only
a vector of samples.</p></td>
    </tr>
    <tr>
      <th>signal_params</th>
      <td><p>A list of additional arguments to pass to <code>signal_fun</code>.</p></td>
    </tr>
    <tr>
      <th>prop_null</th>
      <td><p>The proportion of genes that are null.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>If \(b\) is an effect and \(s\) is an empirical standard deviation, then
we model \(b/s^\alpha\) as being exchangeable.</p></td>
    </tr>
    <tr>
      <th>group_assign</th>
      <td><p>How should we assign groups? Exactly specifying the
proportion of individuals in each group (<code>"frac"</code>), with a
Bernoulli distribution (<code>"random"</code>), or correlated with latent factors
(<code>"cor"</code>)? If <code>group_assign = "cor"</code>, then you have to specify
<code>corvec</code>. If <code>group_assign = "frac"</code> or
<code>group_assign = "random"</code>, then the proportion of samples in each
group is specified with the <code>group_prop</code> argument.</p></td>
    </tr>
    <tr>
      <th>group_prop</th>
      <td><p>The proportion of individuals that are in group 1.
This proportion is deterministic if <code>group_assign = "frac"</code>, and
is the expected proportion if <code>group_assign = "random"</code>. This
argument is not used if <code>group_assign = "cor"</code>.</p></td>
    </tr>
    <tr>
      <th>corvec</th>
      <td><p>A vector of correlations. <code>corvec[i]</code> is the correlation
of the latent group assignment vector with the ith latent confounder.
Only used if <code>group_assign = "cor"</code>. This vector is constrained
so that <code>crossprod(corvec) &lt; 1</code>. The number of latent factors
is taken to be the length of corvec. Note that the correlations of the
latent factors with the observed group-assignment vector (instead of the
latent group-assignment vector) will be <code>corvec * sqrt(2 / pi)</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with the following elements:</p><ul>
<li><p><code>Y</code>: A matrix of altered counts with <code>nsamp</code> rows
       and <code>ngene</code> columns.</p></li>
<li><p><code>X</code>: A design matrix. The first column contains a vector ones (for an
       intercept term) and the second column contains an indicator for group membership.</p></li>
<li><p><code>beta</code>: The approximately true effect sizes of \(log(Y) ~ X\beta\).</p></li>
<li><p><code>corassign</code>: The output from the call to <code><a href='corassign.html'>corassign</a></code>.
       Only returned if <code>group_assign = "cor"</code>.</p></li>
</ul>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Given a matrix of RNA-seq counts, this function will randomly select two groups of
samples and add signal to a known proportion of the genes. This signal
is the log (base 2) effect size of the group indicator in a linear model.
The user may specify the distribution of the effects.</p>
<p>The Poisson thinning approach first randomly assigns samples to be in one of two groups. Then,
given this assignment, will Binomially sample counts with a sample size of the gene expression
counts and a probability that is a function of the effect size. For details, see
Gerard and Stephens (2017).</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Gerard, D., &amp; Stephens, M. (2017).
    Unifying and generalizing methods for removing unwanted variation based
    on negative controls. arXiv preprint arXiv:1705.08393.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
          </ul>

    <h2>Author</h2>
    <p>David Gerard</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

