<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Subsample the rows and columns of a count matrix. — select_counts • seqgendiff</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Subsample the rows and columns of a count matrix. — select_counts" />

<meta property="og:description" content="It is a good idea to subsample (each iteration) the genes and samples from
a real RNA-seq dataset prior to applying thin_diff
(and related functions) so that your conclusions are not dependent on the
specific structure of your dataset. This function is designed to efficiently
do this for you." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/seqgendiff">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Subsample the rows and columns of a count matrix.</h1>
    <small class="dont-index">Source: <a href='https://github.com/dcgerard/seqgendiff/blob/master/R/select_counts.R'><code>R/select_counts.R</code></a></small>
    <div class="hidden name"><code>select_counts.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>It is a good idea to subsample (each iteration) the genes and samples from
a real RNA-seq dataset prior to applying <code><a href='thin_diff.html'>thin_diff</a></code>
(and related functions) so that your conclusions are not dependent on the
specific structure of your dataset. This function is designed to efficiently
do this for you.</p>
    
    </div>

    <pre class="usage"><span class='fu'>select_counts</span>(<span class='no'>mat</span>, <span class='kw'>nsamp</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/nrow'>ncol</a></span>(<span class='no'>mat</span>), <span class='kw'>ngene</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/nrow'>nrow</a></span>(<span class='no'>mat</span>),
  <span class='kw'>gselect</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"random"</span>, <span class='st'>"max"</span>, <span class='st'>"mean_max"</span>, <span class='st'>"custom"</span>), <span class='kw'>gvec</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>filter_first</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>nskip</span> <span class='kw'>=</span> <span class='fl'>0L</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mat</th>
      <td><p>A numeric matrix of RNA-seq counts. The rows index the genes
and the columns index the samples.</p></td>
    </tr>
    <tr>
      <th>nsamp</th>
      <td><p>The number of samples (columns) to select from <code>mat</code>.</p></td>
    </tr>
    <tr>
      <th>ngene</th>
      <td><p>The number of genes (rows) to select from <code>mat</code>.</p></td>
    </tr>
    <tr>
      <th>gselect</th>
      <td><p>How should we select the subset of genes? Options include:</p><dl class='dl-horizontal'>
  <dt><code>random</code></dt><dd><p>Randomly select the genes, with each gene having
      an equal probability of being included in the subsampled matrix.</p></dd>
  <dt><code>max</code></dt><dd><p>Choose the <code>ngene</code> most median-expressed genes.
      Ties are broken by mean-expression.</p></dd>
  <dt><code>mean_max</code></dt><dd><p>Choose the <code>ngene</code> most mean-expressed
      genes.</p></dd>
  <dt><code>custom</code></dt><dd><p>A user-specified list of genes. If
      <code>gselect = "custom"</code> then <code>gvec</code> needs to be
      non-<code>NULL</code>.</p></dd>
</dl></td>
    </tr>
    <tr>
      <th>gvec</th>
      <td><p>A logical vector of length <code><a href='https://www.rdocumentation.org/packages/base/topics/nrow'>nrow(mat)</a></code>. A <code>TRUE</code>
in position \(i\) indicates inclusion into the smaller dataset.
Hence, <code><a href='https://www.rdocumentation.org/packages/base/topics/sum'>sum(gvec)</a></code> should equal <code>ngene</code>.</p></td>
    </tr>
    <tr>
      <th>filter_first</th>
      <td><p>Should we first filter genes by the method of
Chen et al. (2016) (<code>TRUE</code>) or not (<code>FALSE</code>)? If
<code>TRUE</code> then the <code>edgeR</code> package should be installed.</p></td>
    </tr>
    <tr>
      <th>nskip</th>
      <td><p>The number of median-maximally expressed genes to skip.
Not used if <code>gselect = "custom"</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A numeric matrix, which is a <code>ngene</code> by <code>nsamp</code> sub-matrix
    of <code>mat</code>. If <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames(mat)</a></code> is <code>NULL</code>, then the
    row names of the returned matrix are the indices in <code>mat</code> of the
    selected genes. If <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames(mat)</a></code> is <code>NULL</code>, then the
    column names of the returned matrix are the indices in <code>mat</code> of
    the selected samples.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The samples (columns) are chosen randomly, with each sample having
an equal probability of being in the sub-matrix. The genes are selected
according to one of four schemes (see the description of the <code>gselect</code>
argument).</p>
<p>If you have edgeR installed, then some functionality is provided for
filtering out the lowest expressed genes prior to applying subsampling
(see the <code>filter_first</code> argument).
This filtering scheme is described in Chen et al. (2016).
If you want more control over this filtering, you should use
the <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/filterByExpr'>filterByExpr</a></code> function from edgeR directly. You
can install edgeR by following instructions here:
<a href='https://doi.org/doi:10.18129/B9.bioc.edgeR'>https://doi.org/doi:10.18129/B9.bioc.edgeR</a>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <ul>
<li><p>Chen, Yunshun, Aaron TL Lun, and Gordon K. Smyth. "From reads to genes to pathways: differential expression analysis of RNA-Seq experiments using Rsubread and the edgeR quasi-likelihood pipeline." F1000Research 5 (2016).</p></li>
</ul>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Simulate data from given matrix of counts</span>
<span class='co'>## In practice, you would obtain mat from a real dataset, not simulate it.</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Random'>set.seed</a></span>(<span class='fl'>1</span>)
<span class='no'>n</span>   <span class='kw'>&lt;-</span> <span class='fl'>100</span>
<span class='no'>p</span>   <span class='kw'>&lt;-</span> <span class='fl'>1000</span>
<span class='no'>mat</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='kw pkg'>stats</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Poisson'>rpois</a></span>(<span class='no'>n</span> * <span class='no'>p</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='fl'>50</span>), <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='no'>p</span>)

<span class='co'>## Subsample the matrix, then feed it into a thinning function</span>
<span class='no'>submat</span> <span class='kw'>&lt;-</span> <span class='fu'>select_counts</span>(<span class='kw'>mat</span> <span class='kw'>=</span> <span class='no'>mat</span>, <span class='kw'>nsamp</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>ngene</span> <span class='kw'>=</span> <span class='fl'>100</span>)
<span class='no'>thout</span>  <span class='kw'>&lt;-</span> <span class='fu'><a href='thin_2group.html'>thin_2group</a></span>(<span class='kw'>mat</span> <span class='kw'>=</span> <span class='no'>submat</span>, <span class='kw'>prop_null</span> <span class='kw'>=</span> <span class='fl'>0.5</span>)

<span class='co'>## The rownames and colnames (if NULL in mat) tell you which genes/samples</span>
<span class='co'>## were selected.</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames</a></span>(<span class='no'>submat</span>)</div><div class='output co'>#&gt;   [1] "12"  "22"  "40"  "57"  "58"  "70"  "79"  "110" "113" "124" "125" "137"
#&gt;  [13] "141" "142" "153" "186" "200" "236" "237" "239" "251" "254" "257" "259"
#&gt;  [25] "260" "285" "290" "293" "321" "329" "333" "342" "343" "356" "359" "369"
#&gt;  [37] "383" "394" "401" "420" "421" "435" "437" "467" "468" "476" "481" "490"
#&gt;  [49] "492" "496" "507" "519" "520" "549" "588" "597" "616" "620" "641" "647"
#&gt;  [61] "648" "652" "655" "657" "659" "683" "704" "714" "717" "728" "729" "732"
#&gt;  [73] "734" "740" "745" "764" "766" "773" "778" "779" "782" "789" "801" "816"
#&gt;  [85] "836" "845" "848" "849" "857" "863" "877" "904" "913" "919" "930" "947"
#&gt;  [97] "952" "971" "992" "999"</div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames</a></span>(<span class='no'>submat</span>)</div><div class='output co'>#&gt;  [1] "3"  "9"  "13" "14" "40" "80" "84" "86" "94" "99"</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>David Gerard</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

