<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Binomial thinning in the two-group model. — thin_2group • seqgendiff</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Binomial thinning in the two-group model. — thin_2group" />

<meta property="og:description" content="Given a matrix of real RNA-seq counts, this function will
randomly assign samples to one of two groups, draw
the log2-fold change in expression between two groups for each gene,
and add this signal to the RNA-seq counts matrix. The user may specify
the proportion of samples in each group, the proportion of null genes
(where the log2-fold change is 0),
and the signal function. This is a specific application of the
general binomial thinning approach implemented in thin_diff." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/seqgendiff">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Binomial thinning in the two-group model.</h1>
    <small class="dont-index">Source: <a href='https://github.com/dcgerard/seqgendiff/blob/master/R/genthin.R'><code>R/genthin.R</code></a></small>
    <div class="hidden name"><code>thin_2group.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Given a matrix of real RNA-seq counts, this function will
randomly assign samples to one of two groups, draw
the log2-fold change in expression between two groups for each gene,
and add this signal to the RNA-seq counts matrix. The user may specify
the proportion of samples in each group, the proportion of null genes
(where the log2-fold change is 0),
and the signal function. This is a specific application of the
general binomial thinning approach implemented in <code><a href='thin_diff.html'>thin_diff</a></code>.</p>
    
    </div>

    <pre class="usage"><span class='fu'>thin_2group</span>(<span class='no'>mat</span>, <span class='kw'>prop_null</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>signal_fun</span> <span class='kw'>=</span> <span class='kw pkg'>stats</span><span class='kw ns'>::</span><span class='no'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>,
  <span class='kw'>signal_params</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/list'>list</a></span>(<span class='kw'>mean</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>sd</span> <span class='kw'>=</span> <span class='fl'>1</span>), <span class='kw'>group_prop</span> <span class='kw'>=</span> <span class='fl'>0.5</span>,
  <span class='kw'>corvec</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>permute_method</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"optmatch"</span>,
  <span class='st'>"hungarian"</span>, <span class='st'>"marriage"</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mat</th>
      <td><p>A numeric matrix of RNA-seq counts. The rows index the genes and
the columns index the samples.</p></td>
    </tr>
    <tr>
      <th>prop_null</th>
      <td><p>The proportion of genes that are null.</p></td>
    </tr>
    <tr>
      <th>signal_fun</th>
      <td><p>A function that returns the signal. This should take as
input <code>n</code> for the number of samples to return and then return only
a vector of samples. Additional parameters may be passed through
<code>signal_params</code>.</p></td>
    </tr>
    <tr>
      <th>signal_params</th>
      <td><p>A list of additional arguments to pass to
<code>signal_fun</code>.</p></td>
    </tr>
    <tr>
      <th>group_prop</th>
      <td><p>The proportion of individuals that are in group 1.</p></td>
    </tr>
    <tr>
      <th>corvec</th>
      <td><p>A vector of target correlations. <code>corvec[i]</code> is the
target correlation of the latent group assignment vector with the
<code>i</code>th surrogate variable. The default is to set this to <code>NULL</code>,
in which case group assignment is made independently of any
unobserved confounding.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>The scaling factor for the signal distribution from
Stephens (2016). If \(x_1, x_2, ..., x_n\) are drawn from
<code>signal_fun</code>, then the signal is set to
\(x_1 s_1^{\alpha}, x_2 s_2^{\alpha}, ..., x_n s_n^{\alpha}\), where
\(s_g\) is the empirical standard deviation of gene \(g\).
Setting this to <code>0</code> means that the effects are exchangeable, setting
this to <code>1</code> corresponds to the p-value prior of
Wakefield (2009). You would rarely set this to anything but <code>0</code>
or <code>1</code>.</p></td>
    </tr>
    <tr>
      <th>permute_method</th>
      <td><p>Should we use the optimal matching technique from Hansen and
Klopfer (2006) (<code>"optmatch"</code>), the Gale-Shapley algorithm
for stable marriages (<code>"marriage"</code>) (Gale and Shapley, 1962)
as implemented in the matchingR package, or the Hungarian algorithm
(Papadimitriou and Steiglitz, 1982) (<code>"hungarian"</code>)
as implemented in the clue package (Hornik, 2005)?
The <code>"optmatch"</code> method works really well
but does take a lot more computational time if you have, say, 1000
samples. If you use the <code>"optmatch"</code> option, you should note
that the optmatch package uses a super strange license:
<a href='https://cran.r-project.org/web/packages/optmatch/LICENSE'>https://cran.r-project.org/web/packages/optmatch/LICENSE</a>. If this
license doesn't work for you (because you are not in academia, or
because you don't believe in restrictive licenses), then
try out the <code>"hungarian"</code> method.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list-like S3 object of class <code>ThinData</code>.
Components include some or all of the following:</p><dl class='dl-horizontal'>
  <dt><code>mat</code></dt><dd><p>The modified matrix of counts.</p></dd>
  <dt><code>designmat</code></dt><dd><p>The design matrix of variables used to simulate
      signal. This is made by column-binding <code>design_fixed</code> and the
      permuted version of <code>design_perm</code>.</p></dd>
  <dt><code>coefmat</code></dt><dd><p>A matrix of coefficients corresponding to
      <code>designmat</code>.</p></dd>
  <dt><code>design_obs</code></dt><dd><p>Additional variables that should be included in
      your design matrix in downstream fittings. This is made by
      column-binding the vector of 1's with <code>design_obs</code>.</p></dd>
  <dt><code>sv</code></dt><dd><p>A matrix of estimated surrogate variables. In simulation
      studies you would probably leave this out and estimate your own
      surrogate variables.</p></dd>
  <dt><code>cormat</code></dt><dd><p>A matrix of target correlations between the
      surrogate variables and the permuted variables in the design matrix.
      This might be different from the <code>target_cor</code> you input because
      we pass it through <code><a href='fix_cor.html'>fix_cor</a></code> to ensure
      positive semi-definiteness of the resulting covariance matrix.</p></dd>
  <dt><code>matching_var</code></dt><dd><p>A matrix of simulated variables used to
      permute <code>design_perm</code> if the <code>target_cor</code> is not
      <code>NULL</code>.</p></dd>
</dl>

    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <ul>
<li><p>Stephens, Matthew. "False discovery rates: a new deal." Biostatistics 18, no. 2 (2016): 275-294.</p></li>
<li><p>Wakefield, Jon. "Bayes factors for genome-wide association studies: comparison with P-values." Genetic epidemiology 33, no. 1 (2009): 79-86.</p></li>
<li><p>Hansen, Ben B., and Stephanie Olsen Klopfer. "Optimal full matching and related designs via network flows." Journal of computational and Graphical Statistics 15, no. 3 (2006): 609-627.</p></li>
<li><p>Gale, David, and Lloyd S. Shapley. "College admissions and the stability of marriage." The American Mathematical Monthly 69, no. 1 (1962): 9-15.</p></li>
<li><p>C. Papadimitriou and K. Steiglitz (1982), Combinatorial Optimization: Algorithms and Complexity. Englewood Cliffs: Prentice Hall.</p></li>
<li><p>Hornik K (2005). "A CLUE for CLUster Ensembles." Journal of Statistical Software, 14(12). doi: 10.18637/jss.v014.i12</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><dl class='dl-horizontal'>
  <dt><code><a href='select_counts.html'>select_counts</a></code></dt><dd><p>For subsampling the rows and columns
      of your real RNA-seq count matrix prior to applying binomial thinning.</p></dd>
  <dt><code><a href='thin_diff.html'>thin_diff</a></code></dt><dd><p>For the more general thinning approach.</p></dd>
  <dt><code><a href='ThinDataToSummarizedExperiment.html'>ThinDataToSummarizedExperiment</a></code></dt><dd><p>For converting a
      ThinData object to a SummarizedExperiment object.</p></dd>
  <dt><code><a href='ThinDataToDESeqDataSet.html'>ThinDataToDESeqDataSet</a></code></dt><dd><p>For converting a
      ThinData object to a DESeqDataSet object.</p></dd>
</dl></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Simulate data from given matrix of counts</span>
<span class='co'>## In practice, you would obtain Y from a real dataset, not simulate it.</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Random'>set.seed</a></span>(<span class='fl'>1</span>)
<span class='no'>nsamp</span> <span class='kw'>&lt;-</span> <span class='fl'>10</span>
<span class='no'>ngene</span> <span class='kw'>&lt;-</span> <span class='fl'>1000</span>
<span class='no'>Y</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='kw pkg'>stats</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Poisson'>rpois</a></span>(<span class='no'>nsamp</span> * <span class='no'>ngene</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='fl'>50</span>), <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='no'>ngene</span>)
<span class='no'>thinout</span> <span class='kw'>&lt;-</span> <span class='fu'>thin_2group</span>(<span class='kw'>mat</span>           <span class='kw'>=</span> <span class='no'>Y</span>,
                       <span class='kw'>prop_null</span>     <span class='kw'>=</span> <span class='fl'>0.9</span>,
                       <span class='kw'>signal_fun</span>    <span class='kw'>=</span> <span class='kw pkg'>stats</span><span class='kw ns'>::</span><span class='no'><a href='https://www.rdocumentation.org/packages/stats/topics/Exponential'>rexp</a></span>,
                       <span class='kw'>signal_params</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/list'>list</a></span>(<span class='kw'>rate</span> <span class='kw'>=</span> <span class='fl'>0.5</span>))

<span class='co'>## 90 percent of genes are null</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/mean'>mean</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/MathFun'>abs</a></span>(<span class='no'>thinout</span>$<span class='no'>coef</span>) <span class='kw'>&lt;</span> <span class='fl'>10</span>^-<span class='fl'>6</span>)</div><div class='output co'>#&gt; [1] 0.9</div><div class='input'>
<span class='co'>## Check the estimates of the log2-fold change</span>
<span class='no'>Ynew</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Log'>log2</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/t'>t</a></span>(<span class='no'>thinout</span>$<span class='no'>mat</span> + <span class='fl'>0.5</span>))
<span class='no'>X</span>    <span class='kw'>&lt;-</span> <span class='no'>thinout</span>$<span class='no'>designmat</span>
<span class='no'>Bhat</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/coef'>coef</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/lm'>lm</a></span>(<span class='no'>Ynew</span> ~ <span class='no'>X</span>))[<span class='st'>"X"</span>, ]
<span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/plot'>plot</a></span>(<span class='no'>thinout</span>$<span class='no'>coefmat</span>,
     <span class='no'>Bhat</span>,
     <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"log2-fold change"</span>,
     <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"Estimated log2-fold change"</span>)</div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/abline'>abline</a></span>(<span class='fl'>0</span>, <span class='fl'>1</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>lwd</span> <span class='kw'>=</span> <span class='fl'>2</span>)</div><div class='img'><img src='thin_2group-1.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>David Gerard</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

