<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Poisson thinning for differential expression analysis. — thin_diff • seqgendiff</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Poisson thinning for differential expression analysis. — thin_diff" />

<meta property="og:description" content="Given a matrix of real RNA-seq counts, this function will add a known
amount of signal to the count matrix. This signal is given in the form
of a Poisson generalized linear model with a log (base 2) link. The user may
specify any arbitrary design matrix and coefficient matrix. The user
may also control for the amount of correlation between the observed
covariates and any unobserved surrogate variables." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/seqgendiff">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Poisson thinning for differential expression analysis.</h1>
    <small class="dont-index">Source: <a href='https://github.com/dcgerard/seqgendiff/blob/master/R/genthin.R'><code>R/genthin.R</code></a></small>
    <div class="hidden name"><code>thin_diff.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Given a matrix of real RNA-seq counts, this function will add a known
amount of signal to the count matrix. This signal is given in the form
of a Poisson generalized linear model with a log (base 2) link. The user may
specify any arbitrary design matrix and coefficient matrix. The user
may also control for the amount of correlation between the observed
covariates and any unobserved surrogate variables.</p>
    
    </div>

    <pre class="usage"><span class='fu'>thin_diff</span>(<span class='no'>mat</span>, <span class='kw'>design_fixed</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>coef_fixed</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>design_perm</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>coef_perm</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>target_cor</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>use_sva</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>design_obs</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>relative</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mat</th>
      <td><p>A numeric matrix of counts. The rows index the genes and the
columns index the samples (as is usual in RNA-seq).</p></td>
    </tr>
    <tr>
      <th>design_fixed</th>
      <td><p>A numeric design matrix whose rows are fixed and not permuted.
The rows index the samples and the columns index the variables.
The intercept should <em>not</em> be included.</p></td>
    </tr>
    <tr>
      <th>coef_fixed</th>
      <td><p>A numeric matrix. The coefficients corresponding to <code>design_fixed</code>.
The rows index the genes and the columns index the variables.</p></td>
    </tr>
    <tr>
      <th>design_perm</th>
      <td><p>A numeric design matrix whose rows are to be permuted (thus
controlling the amount by which they are correlated with the surrogate variables).
The rows index the samples and the columns index the variables.
The intercept should <em>not</em> be included.</p></td>
    </tr>
    <tr>
      <th>coef_perm</th>
      <td><p>A numeric matrix. The coefficients corresponding to <code>design_perm</code>.
The rows index the genes and the columns index the variables.</p></td>
    </tr>
    <tr>
      <th>target_cor</th>
      <td><p>A numeric matrix of target correlations betweens the variables in
<code>design_perm</code> and the surrogate variables. The rows index the
observed covariates and the columns index the surrogate variables.
The number of columns indicates the number of surrogate variables.
Set this to <code>NULL</code> to indicate uncorrelated.</p></td>
    </tr>
    <tr>
      <th>use_sva</th>
      <td><p>A logical. Should we use SVA using
<code>design_obs</code> to estimate the hidden covariates (<code>TRUE</code>)
or should we just do an SVD on <code><a href='https://www.rdocumentation.org/packages/base/topics/Log'>log2(mat + 0.5)</a></code> after
regressing out <code>design_obs</code> (<code>FALSE</code>)? Setting this to
<code>TRUE</code> allows the surrogate variables to be correlated with the
observed covariates, while setting this to <code>FALSE</code> assumes that
the surrogate variabes are orthogonal to the observed covariates. This
option only matters if <code>design_obs</code> is not <code>NULL</code>.
Defaults to <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>design_obs</th>
      <td><p>A numeric matrix of observed covariates that are NOT to be a
part of the signal generating process. Only used in estimating the
surrogate variables (if <code>target_cor</code> is not <code>NULL</code>).
The intercept should <em>not</em> be included.</p></td>
    </tr>
    <tr>
      <th>relative</th>
      <td><p>A logical. Should we apply relative thinning (<code>TRUE</code>)
or absolute thinning (<code>FALSE</code>). Only experts should change
the default.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An S3 object of class <code>ThinData</code>. Components include some or
all of the following:</p><dl class='dl-horizontal'>
  <dt><code>mat</code></dt><dd><p>The modified matrix of counts.</p></dd>
  <dt><code>designmat</code></dt><dd><p>The design matrix, excluding an intercept term.</p></dd>
  <dt><code>coefmat</code></dt><dd><p>A matrix of coefficients corresponding to <code>designmat</code></p></dd>
  <dt><code>sv</code></dt><dd><p>A matrix of estimated surrogate variables.</p></dd>
  <dt><code>cormat</code></dt><dd><p>A matrix of target correlations between the
      surrogate variables and the permuted variables in the design matrix.</p></dd>
  <dt><code>matching_var</code></dt><dd><p>A matrix of simulated variables used to
      permute the permuted components of the design matrix.</p></dd>
</dl>

    
    <h2 class="hasAnchor" id="mathematical-formulation"><a class="anchor" href="#mathematical-formulation"></a>Mathematical Formulation</h2>

    
    <p>Let</p><dl class='dl-horizontal'>
  <dt>\(N\)</dt><dd><p>Be the number of samples.</p></dd>
  <dt>\(G\)</dt><dd><p>Be the number of genes.</p></dd>
  <dt>\(Y\)</dt><dd><p>Be an \(G\) by \(N\) matrix of real RNA-seq counts.
      This is <code>mat</code>.</p></dd>
  <dt>\(X_1\)</dt><dd><p>Be an \(N\) by \(P_1\) user-provided design matrix.
      This is <code>design_fixed</code>.</p></dd>
  <dt>\(X_2\)</dt><dd><p>Be an \(N\) by \(P_2\) user-provided design matrix.
      This is <code>design_perm</code>.</p></dd>
  <dt>\(X_3\)</dt><dd><p>Be an \(N\) by \(P_3\) matrix of known covariates.
      This is <code>design_obs</code>.</p></dd>
  <dt>\(Z\)</dt><dd><p>Be an \(N\) by \(K\) matrix of unobserved surrogate
       variables. This is estimated when <code>target_cor</code> is not
       <code>NULL</code>.</p></dd>
</dl><p>We assume that \(Y\) is Poisson distributed given \(X_3\) and
\(Z\) such that
$$\log_2(EY) = 1_N\mu' + B_3X_3' + AZ'.$$
<code>thin_diff()</code> will take as inpute \(X_1\), \(X_2\), \(B_1\),
\(B_2\), and will output a \(\tilde{Y}\) and \(W\) such that
\(\tilde{Y}\) is Poisson distributed given \(X_1\), \(X_2\), \(X_3\),
\(W\), and \(Z\) such that
$$\log_2(E\tilde{Y}) \approx 1_N\tilde{\mu}' + B_1X_1' + B_2X_2'W' + B_3X_3' + AZ',$$
where \(W\) is an \(N\) by \(N\) permutation matrix. \(W\) is randomly
drawn so that \(X_2\) and \(Z\) are correlated approximately according
to a target correlation matrix.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><dl class='dl-horizontal'>
  <dt><code><a href='thin_2group.html'>thin_2group</a></code></dt><dd><p>For the specific application of
      <code>thin_diff</code> to the two-group model.</p></dd>
  <dt><code><a href='thin_lib.html'>thin_lib</a></code></dt><dd><p>For the specific application of
      <code>thin_diff</code> to library size thinning.</p></dd>
  <dt><code><a href='thin_base.html'>thin_base</a></code></dt><dd><p>For the underlying thinning function
      used in <code>thin_diff</code>.</p></dd>
</dl></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Generate simulated data with surrogate variables</span>
<span class='co'>## In practice, you would obtain mat from a real dataset, not simulate it.</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Random'>set.seed</a></span>(<span class='fl'>1</span>)
<span class='no'>n</span> <span class='kw'>&lt;-</span> <span class='fl'>10</span>
<span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fl'>1000</span>
<span class='no'>Z</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/MathFun'>abs</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>(<span class='no'>n</span>, <span class='kw'>sd</span> <span class='kw'>=</span> <span class='fl'>4</span>)))
<span class='no'>alpha</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/MathFun'>abs</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>(<span class='no'>p</span>, <span class='kw'>sd</span> <span class='kw'>=</span> <span class='fl'>1</span>)))
<span class='no'>mat</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Round'>round</a></span>(<span class='fl'>2</span>^(<span class='no'>alpha</span> <span class='kw'>%*%</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/t'>t</a></span>(<span class='no'>Z</span>) + <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/MathFun'>abs</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>(<span class='no'>n</span> * <span class='no'>p</span>, <span class='kw'>sd</span> <span class='kw'>=</span> <span class='fl'>5</span>),
                                            <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='no'>p</span>,
                                            <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='no'>n</span>))))

<span class='co'>## Choose simulation parameters</span>
<span class='no'>design_perm</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/cbind'>cbind</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>0</span>, <span class='fl'>1</span>), <span class='kw'>length.out</span> <span class='kw'>=</span> <span class='no'>n</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Uniform'>runif</a></span>(<span class='no'>n</span>))
<span class='no'>coef_perm</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>(<span class='no'>p</span> * <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/nrow'>ncol</a></span>(<span class='no'>design_perm</span>), <span class='kw'>sd</span> <span class='kw'>=</span> <span class='fl'>6</span>), <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='no'>p</span>)

<span class='co'>## Specify one surrogate variable (number of columns in taget_cor),</span>
<span class='co'>## highly correlated with first observed covariate and uncorrelated</span>
<span class='co'>## with second observed covariate</span>
<span class='no'>target_cor</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>0.9</span>, <span class='fl'>0</span>))

<span class='co'>## Thin</span>
<span class='no'>thout</span> <span class='kw'>&lt;-</span> <span class='fu'>thin_diff</span>(<span class='kw'>mat</span> <span class='kw'>=</span> <span class='no'>mat</span>,
                   <span class='kw'>design_perm</span> <span class='kw'>=</span> <span class='no'>design_perm</span>,
                   <span class='kw'>coef_perm</span> <span class='kw'>=</span> <span class='no'>coef_perm</span>,
                   <span class='kw'>target_cor</span> <span class='kw'>=</span> <span class='no'>target_cor</span>)

<span class='co'>## target_cor approximates correlation between estimated surrogate variable</span>
<span class='co'>## and matching variable.</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/cor'>cor</a></span>(<span class='no'>thout</span>$<span class='no'>matching_var</span>, <span class='no'>thout</span>$<span class='no'>sv</span>)</div><div class='output co'>#&gt;            [,1]
#&gt; [1,] 0.92652364
#&gt; [2,] 0.08621485</div><div class='input'>
<span class='co'>## Estimated surrogate variable is associated with true surrogate variable</span>
<span class='co'>## (because the signal is strong in this case)</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/plot'>plot</a></span>(<span class='no'>Z</span>, <span class='no'>thout</span>$<span class='no'>sv</span>, <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"True SV"</span>, <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"Estimated SV"</span>)</div><div class='img'><img src='thin_diff-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'>## So target_cor approximates correlation between surrogate variable and</span>
<span class='co'>## matching variables</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/cor'>cor</a></span>(<span class='no'>thout</span>$<span class='no'>matching_var</span>, <span class='no'>Z</span>)</div><div class='output co'>#&gt;            [,1]
#&gt; [1,] 0.91827513
#&gt; [2,] 0.08834679</div><div class='input'>
<span class='co'>## Correlation between permuted covariates and surrogate variables are less</span>
<span class='co'>## close to target_cor</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/cor'>cor</a></span>(<span class='no'>thout</span>$<span class='no'>designmat</span>, <span class='no'>Z</span>)</div><div class='output co'>#&gt;            [,1]
#&gt; [1,]  0.5890975
#&gt; [2,] -0.2183653</div><div class='input'>
<span class='co'>## Estimated signal is correlated to true single. First variable is slightly</span>
<span class='co'>## biased because the surrogate variable is not included.</span>
<span class='no'>Ynew</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Log'>log2</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/t'>t</a></span>(<span class='no'>thout</span>$<span class='no'>mat</span>) + <span class='fl'>0.5</span>)
<span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='no'>thout</span>$<span class='no'>designmat</span>
<span class='no'>coef_est</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/t'>t</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/coef'>coef</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/lm'>lm</a></span>(<span class='no'>Ynew</span> ~ <span class='no'>X</span>))[<span class='fl'>2</span>:<span class='fl'>3</span>, ])

<span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/plot'>plot</a></span>(<span class='no'>thout</span>$<span class='no'>coefmat</span>[, <span class='fl'>1</span>], <span class='no'>coef_est</span>[, <span class='fl'>1</span>])</div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/abline'>abline</a></span>(<span class='fl'>0</span>, <span class='fl'>1</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>lwd</span> <span class='kw'>=</span> <span class='fl'>2</span>)</div><div class='img'><img src='thin_diff-2.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/plot'>plot</a></span>(<span class='no'>thout</span>$<span class='no'>coefmat</span>[, <span class='fl'>2</span>], <span class='no'>coef_est</span>[, <span class='fl'>2</span>])</div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/abline'>abline</a></span>(<span class='fl'>0</span>, <span class='fl'>1</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>lwd</span> <span class='kw'>=</span> <span class='fl'>2</span>)</div><div class='img'><img src='thin_diff-3.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'>## But estimated coefficient of the first variable is slightly closer when</span>
<span class='co'>## the surrogate variable is included.</span>
<span class='no'>Ynew</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Log'>log2</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/t'>t</a></span>(<span class='no'>thout</span>$<span class='no'>mat</span>) + <span class='fl'>0.5</span>)
<span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/cbind'>cbind</a></span>(<span class='no'>thout</span>$<span class='no'>designmat</span>, <span class='no'>thout</span>$<span class='no'>sv</span>)
<span class='no'>coef_est</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/t'>t</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/coef'>coef</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/lm'>lm</a></span>(<span class='no'>Ynew</span> ~ <span class='no'>X</span>))[<span class='fl'>2</span>:<span class='fl'>3</span>, ])

<span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/plot'>plot</a></span>(<span class='no'>thout</span>$<span class='no'>coefmat</span>[, <span class='fl'>1</span>], <span class='no'>coef_est</span>[, <span class='fl'>1</span>])</div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/abline'>abline</a></span>(<span class='fl'>0</span>, <span class='fl'>1</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>lwd</span> <span class='kw'>=</span> <span class='fl'>2</span>)</div><div class='img'><img src='thin_diff-4.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/plot'>plot</a></span>(<span class='no'>thout</span>$<span class='no'>coefmat</span>[, <span class='fl'>2</span>], <span class='no'>coef_est</span>[, <span class='fl'>2</span>])</div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/graphics/topics/abline'>abline</a></span>(<span class='fl'>0</span>, <span class='fl'>1</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>lwd</span> <span class='kw'>=</span> <span class='fl'>2</span>)</div><div class='img'><img src='thin_diff-5.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#mathematical-formulation">Mathematical Formulation</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>David Gerard</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

