<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Binomial thinning for altering total gene expression levels — thin_gene • seqgendiff</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Binomial thinning for altering total gene expression levels — thin_gene" />
<meta property="og:description" content="Given a matrix of real RNA-seq counts, this function will apply a
separate, user-provided thinning factor to each gene. This uniformly
lowers the counts for all samples in a gene. The thinning factor
should be provided on the log2-scale. This is a specific application
of the binomial thinning approach in thin_diff. The method is
described in detail in Gerard (2019)." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqgendiff</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/different_thinners.html">Applying Different Thinning Functions</a>
    </li>
    <li>
      <a href="../articles/simulate_rnaseq.html">Simulate RNA-seq Data from Real Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/seqgendiff">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Binomial thinning for altering total gene expression levels</h1>
    <small class="dont-index">Source: <a href='https://github.com/dcgerard/seqgendiff/blob/master/R/genthin.R'><code>R/genthin.R</code></a></small>
    <div class="hidden name"><code>thin_gene.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a matrix of real RNA-seq counts, this function will apply a
separate, user-provided thinning factor to each gene. This uniformly
lowers the counts for all samples in a gene. The thinning factor
should be provided on the log2-scale. This is a specific application
of the binomial thinning approach in <code><a href='thin_diff.html'>thin_diff</a></code>. The method is
described in detail in Gerard (2019).</p>
    </div>

    <pre class="usage"><span class='fu'>thin_gene</span>(<span class='no'>mat</span>, <span class='no'>thinlog2</span>, <span class='kw'>relative</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mat</th>
      <td><p>A numeric matrix of RNA-seq counts. The rows index the genes and
the columns index the samples.</p></td>
    </tr>
    <tr>
      <th>thinlog2</th>
      <td><p>A vector of numerics. Element i is the amount to thin
(on the log2 scale) for gene i. For
example, a value of 0 means that we do not thin, a value of 1 means
that we thin by a factor of 2, a value of 2 means we thin by a factor
of 4, etc.</p></td>
    </tr>
    <tr>
      <th>relative</th>
      <td><p>A logical. Should we apply relative thinning (<code>TRUE</code>)
or absolute thinning (<code>FALSE</code>). Only experts should change
the default.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list-like S3 object of class <code>ThinData</code>.
Components include some or all of the following:</p><dl class='dl-horizontal'>
  <dt><code>mat</code></dt><dd><p>The modified matrix of counts.</p></dd>
  <dt><code>designmat</code></dt><dd><p>The design matrix of variables used to simulate
      signal. This is made by column-binding <code>design_fixed</code> and the
      permuted version of <code>design_perm</code>.</p></dd>
  <dt><code>coefmat</code></dt><dd><p>A matrix of coefficients corresponding to
      <code>designmat</code>.</p></dd>
  <dt><code>design_obs</code></dt><dd><p>Additional variables that should be included in
      your design matrix in downstream fittings. This is made by
      column-binding the vector of 1's with <code>design_obs</code>.</p></dd>
  <dt><code>sv</code></dt><dd><p>A matrix of estimated surrogate variables. In simulation
      studies you would probably leave this out and estimate your own
      surrogate variables.</p></dd>
  <dt><code>cormat</code></dt><dd><p>A matrix of target correlations between the
      surrogate variables and the permuted variables in the design matrix.
      This might be different from the <code>target_cor</code> you input because
      we pass it through <code><a href='fix_cor.html'>fix_cor</a></code> to ensure
      positive semi-definiteness of the resulting covariance matrix.</p></dd>
  <dt><code>matching_var</code></dt><dd><p>A matrix of simulated variables used to
      permute <code>design_perm</code> if the <code>target_cor</code> is not
      <code>NULL</code>.</p></dd>

</dl>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    
<ul>
<li><p>Gerard D (2019). "Data-based RNA-seq Simulations by Binomial Thinning." <em>bioRxiv</em>. doi: <a href='https://doi.org/10.1101/758524'>10.1101/758524</a>.</p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<dl class='dl-horizontal'>
  <dt><code><a href='select_counts.html'>select_counts</a></code></dt><dd><p>For subsampling the rows and columns
      of your real RNA-seq count matrix prior to applying binomial thinning.</p></dd>
  <dt><code><a href='thin_diff.html'>thin_diff</a></code></dt><dd><p>For the more general thinning approach.</p></dd>
  <dt><code><a href='thin_lib.html'>thin_lib</a></code></dt><dd><p>For thinning sample-wise instead of
      gene-wise.</p></dd>
  <dt><code><a href='thin_all.html'>thin_all</a></code></dt><dd><p>For thinning all counts uniformly.</p></dd>
  <dt><code><a href='ThinDataToSummarizedExperiment.html'>ThinDataToSummarizedExperiment</a></code></dt><dd><p>For converting a
      ThinData object to a SummarizedExperiment object.</p></dd>
  <dt><code><a href='ThinDataToDESeqDataSet.html'>ThinDataToDESeqDataSet</a></code></dt><dd><p>For converting a
      ThinData object to a DESeqDataSet object.</p></dd>

</dl>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Generate count data and thinning factors</span>
<span class='co'>## In practice, you would obtain mat from a real dataset, not simulate it.</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span>(<span class='fl'>1</span>)
<span class='no'>n</span> <span class='kw'>&lt;-</span> <span class='fl'>10</span>
<span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fl'>1000</span>
<span class='no'>lambda</span> <span class='kw'>&lt;-</span> <span class='fl'>1000</span>
<span class='no'>mat</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span>(<span class='no'>lambda</span>, <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='no'>n</span>, <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='no'>p</span>)
<span class='no'>thinlog2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/Exponential.html'>rexp</a></span>(<span class='kw'>n</span> <span class='kw'>=</span> <span class='no'>p</span>, <span class='kw'>rate</span> <span class='kw'>=</span> <span class='fl'>1</span>)

<span class='co'>## Thin total gene expressions</span>
<span class='no'>thout</span> <span class='kw'>&lt;-</span> <span class='fu'>thin_gene</span>(<span class='kw'>mat</span> <span class='kw'>=</span> <span class='no'>mat</span>, <span class='kw'>thinlog2</span> <span class='kw'>=</span> <span class='no'>thinlog2</span>)

<span class='co'>## Compare empirical thinning proportions to specified thinning proportions</span>
<span class='no'>empirical_propvec</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowMeans</a></span>(<span class='no'>thout</span>$<span class='no'>mat</span>) / <span class='no'>lambda</span>
<span class='no'>specified_propvec</span> <span class='kw'>&lt;-</span> <span class='fl'>2</span> ^ (-<span class='no'>thinlog2</span>)
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.html'>plot</a></span>(<span class='no'>empirical_propvec</span>, <span class='no'>specified_propvec</span>,
     <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"Empirical Thinning Proportion"</span>,
     <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"Specified Thinning Proportion"</span>)</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span>(<span class='fl'>0</span>, <span class='fl'>1</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>lwd</span> <span class='kw'>=</span> <span class='fl'>2</span>)</div><div class='img'><img src='thin_gene-1.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>David Gerard</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


